neutrinordp/xrdp-neutrinordp.c

```diff
@@ -94,10 +94,107 @@ lxrdp_start(struct mod *mod, int w, int h, int bpp)
 
 /******************************************************************************/
 /* return error */
+static void
+set_keyboard_overrides(struct mod *mod)
+{
+    const struct kbd_overrides *ko = &mod->kbd_overrides;
+    rdpSettings *settings  = settings = mod->inst->settings;
+
+    if (mod->allow_client_kbd_settings)
+    {
+        settings->kbd_type = mod->client_info.keyboard_type;
+        settings->kbd_subtype = mod->client_info.keyboard_subtype;
+        /* Define the most common number of function keys, 12.
+           because we can't get it from client. */
+        settings->kbd_fn_keys = 12;
+
+        settings->kbd_layout = mod->client_info.keylayout;
+        if (mod->client_info.keyboard_type == 0x00)
+        {
+            LOG(LOG_LEVEL_WARNING, "keyboard_type:[0x%02x] ,Set on Server",
+                mod->client_info.keyboard_type  );
+        }
+        else if (mod->client_info.keyboard_type == 0x04)
+        {
+            /* When the keyboard type is 0x04 (PC101/102 keyboard) and the lower 16
+               bits are a value of 0x0411(Japanese KeyLayout), the kbd_layout is set
+               to 0x00000411 no matter what the upper 16 bits are. */
+            if ((settings->kbd_layout & 0x0000FFFF) == 0x00000411)
+            {
+                settings->kbd_layout = 0x00000411;
+            }
```

Comment:
Here is a supplement about this code.

I also tested it on Japanese Windows10 with an English keyboard(`kbd_type:0x04`) connected.
For some reason, the 'keyboardLayout' was '0xE0010411'.

I'm confused.

I've added some hardcoding to make sure that `0x00000411` can still be sent to remote Windows in this case.